### PR TITLE
Enable waiting for response for #publish_message

### PR DESCRIPTION
**Pull request by bzwei**
If the `client.publish_message` has a block, consider the message requires a response from the subscriber and the block will be executed upon receiving the response.

The typical usage is
```
  client.publish_message(options) do |result|
    # run this block when receiving a response from the subscriber. Expect result to be an empty string
    # if no execution result is needed.
    # the result does not need to be a string, can be a Hash for example.
  end
```
It make no sense to add similar feature to `#publish_topic` or `#publish_messages`.

### CODE REVIEW

**Comment by bzwei** [comment]
@mkanoor @kbrock

File: lib/manageiq/messaging/background_job.rb [hunk 1]
```diff
@@ -43,11 +46,13 @@ def self.run_job(options)
         obj = obj.find(instance_id) if instance_id
 
         msg_timeout = 600 # TODO: configurable per message
+        result = nil
         Timeout.timeout(msg_timeout) do
-          obj.send(options[:method_name], *args)
+          result = obj.send(options[:method_name], *args)
```
Comment (Fryguy) [inline lib/manageiq/messaging/background_job.rb]: I *think* you can just set result to the block directly

```ruby
result = Timeout.timeout(msg_timeout) do
  obj.send(options[:method_name], *args)
end
```

File: lib/manageiq/messaging/common.rb [hunk 2]
```diff
@@ -74,6 +74,28 @@ def decode_body(headers, raw_body)
...
+        def receive_response(client, service, correlation_id)
+          response_options = {
+            :service  => "#{service}.response",
+            :affinity => correlation_id
+          }
+          queue_name, response_headers = queue_for_subscribe(response_options)
+          client.subscribe(queue_name, response_headers) do |msg|
+            client.ack(msg)
+            yield decode_body(msg.headers, msg.body)
+            client.unsubscribe(queue_name)
```
Comment (Fryguy) [inline lib/manageiq/messaging/common.rb]: Should this be in an ensure block?

File: lib/manageiq/messaging/queue.rb [hunk 3]
```diff
@@ -3,15 +3,20 @@ module Messaging
...
 
-      def self.publish(client, options)
+      def self.publish(client, options, &block)
         assert_options(options, [:message, :service])
 
         address, headers = queue_for_publish(options)
         headers[:sender] = options[:sender] if options[:sender]
         headers[:message_type] = options[:message] if options[:message]
         headers[:class_name] = options[:class_name] if options[:class_name]
+        headers[:correlation_id] = Time.now.to_i.to_s if block_given?
```
Comment (Fryguy) [inline lib/manageiq/messaging/queue.rb]: Not a fan of the name `correlation_id` (in particular the `id` bit), but I can't think of a better name.

File: lib/manageiq/messaging/common.rb (hunk 2 above)
Comment (bzwei) [inline lib/manageiq/messaging/common.rb]: yes

File: lib/manageiq/messaging/queue.rb (hunk 3 above)
Comment (bzwei) [inline lib/manageiq/messaging/queue.rb]: `correlation_id` is a term used by ActiveMQ and Artemis

Comment (bzwei) [inline lib/manageiq/messaging/queue.rb]: Changed `correlation_id` to `correlation_ref` where possible. Here the keyword is reserved by ActiveMQ.

Comment (Fryguy) [inline lib/manageiq/messaging/queue.rb]: Yeah, I like `ref` as a suffix

**Pull request merged**